File: set1/5/main.cpp
#include<iostream>
#include<queue>
#include<algorithm>

using namespace std;

const int dy[] = {-1,1,0,0};
const int dx[] = {0,0,-1,1};
int Map[2005][2005], ans;
int trace[2005][2005];
int N, M;

int changeDir(int value, int curDir){
    if(value == 1) {
        if(curDir == 0 || curDir == 1) return curDir;
        return 5 - curDir;
    }
    if(value == 2) {
        if(curDir == 0 || curDir == 1) return 1 - curDir;
        return curDir;
    }
    if(value == 3) return 3 - curDir;
    if(value == 4) {
        if(curDir == 1 || curDir == 3)return 4 - curDir;
        return 2 - curDir;
    }
    return curDir;
}

bool range(int y, int x) {
    return 0 <= y && y < N && 0 <= x && x < M;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> N >> M;
    queue<pair<int, int>> q;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            cin >> Map[i][j];
            if(Map[i][j] == 9)q.emplace(i, j);
        }
    }

    while(!q.empty()){
        auto [y, x] = q.front(); q.pop();
        if(~trace[y][x] & 16) {
            trace[y][x] |= 16;
            ans++;
        }
        for (int i = 0; i < 4; i++) {
            int dir = i;
            int qy = y, qx = x;
            if(trace[qy][qx] & (1 << dir))continue;
            trace[qy][qx] |= (1 << dir);
            while(true){
                qy += dy[dir]; qx += dx[dir];
                if(!range(qy, qx) || trace[qy][qx] & (1 << dir)) break;
                trace[qy][qx] |= (1 << dir);
                if(~trace[qy][qx] & 16){
                    trace[qy][qx] |= 16;
                    ans++;
                }
                dir = changeDir(Map[qy][qx], dir);
            }
        }
    }
    cout << ans;

    return 0;
}

File: set1/5/main.py
import sys
from collections import deque

def input():
    return sys.stdin.readline().rstrip()

N, M = map(int, input().split())
Map = [list(map(int, input().split())) for i in range(N)]
Q = deque([(i, j) for i in range(N) for j in range(M) if Map[i][j] == 9])

dy, dx = (-1,1,0,0),(0,0,-1,1)

def changedir(x, dir):
    if x == 1: return dir if dir in [0, 1] else 5 - dir
    if x == 2: return 1 - dir if dir in [0, 1] else dir
    if x == 3: return 3 - dir
    if x == 4: return 4 - dir if dir in [1, 3] else 2 - dir
    return dir

def inrange(y, x):
    return 0 <= y < N and 0 <= x < M

ans = 0
trace = [[0 for j in range(M)] for i in range(N)]
while Q:
    y, x = Q.popleft()
    if ~trace[y][x] & 16:
        trace[y][x] |= 16
        ans += 1
    for i in range(4):
        dir = i
        qy, qx = y, x
        if trace[qy][qx] & (1 << dir): continue
        trace[qy][qx] |= (1 << dir)
        while True:
            qy, qx = qy + dy[dir], qx + dx[dir]
            if not inrange(qy, qx) or trace[qy][qx] & (1 << dir): break
            trace[qy][qx] |= (1 << dir)
            if ~trace[qy][qx] & 16:
                trace[qy][qx] |= 16
                ans += 1
            dir = changedir(Map[qy][qx], dir)
print(ans)